core/src/main/java/org/geysermc/geyser/translator/text/MessageTranslator.java

```diff
@@ -428,15 +433,95 @@ public static String normalizeSpace(String string) {
+        Component parsed = componentFromNbtTag(description);
+        return convertMessage(session, parsed);
+    }
+
+    public static Component componentFromNbtTag(Object nbtTag) {
+        return componentFromNbtTag(nbtTag, Style.empty());
+    }
+
+    private static Component componentFromNbtTag(Object nbtTag, Style style) {
+        if (nbtTag instanceof String literal) {
+            return Component.text(literal).style(style);
+        } else if (nbtTag instanceof List<?> list) {
+            return Component.join(JoinConfiguration.builder(), componentsFromNbtList(list, style));
+        } else if (nbtTag instanceof NbtMap map) {
+            Component component = Component.empty();
+            if (map.containsKey("text")) {
+                component = Component.text(map.getString("text"));
+            } else if (map.containsKey("translate")) {
+                String key = map.getString("translate");
+                String fallback = map.getString("fallback", "");
+                List<Component> args = new ArrayList<>();
+
+                if (map.containsKey("with")) {
+                    Object with = map.get("with");
+                    if (with instanceof List<?> list) {
+                        args = componentsFromNbtList(list, style);
+                    } else {
+                        args.add(componentFromNbtTag(with, style));
+                    }
+                }
+                component = Component.translatable(key, fallback, args);
+            }
+
+            Style newStyle = getStyleFromNbtMap(map, style);
+            component = component.style(newStyle);
+
+            Object extra = map.get("extra");
+            if (extra != null) {
+                component = component.append(componentFromNbtTag(extra, newStyle));
+            }
+
+            return component;
+        }
+
+        throw new IllegalArgumentException("Expected tag to be a literal string, a list of components, or a component");
+    }
+
+    private static List<Component> componentsFromNbtList(List<?> list, Style style) {
+        List<Component> components = new ArrayList<>();
+        for (Object entry : list) {
+            components.add(componentFromNbtTag(entry, style));
         }
-        return translate;
+        return components;
+    }
+
+    public static Style getStyleFromNbtMap(NbtMap map) {
+        return getStyleFromNbtMap(map, Style.empty());
+    }
+
+    public static Style getStyleFromNbtMap(NbtMap map, Style base) {
+        Style.Builder newStyle = Style.style().merge(base);
+
+        if (map.containsKey("color", NbtType.STRING)) {
+            String colorString = map.getString("color");
+            if (colorString.startsWith(TextColor.HEX_PREFIX)) {
+                newStyle.color(TextColor.fromHexString(colorString));
+            } else {
+                newStyle.color(NamedTextColor.NAMES.value(colorString));
+            }
+        }
+        if (map.containsKey("bold", NbtType.BYTE)) {
```

Comment:
I see what you're going for here... Would Cloudburst's listenForBoolean suffice instead?

Reply:
Thanks! I'm not too familiar with Cloudburst's library and wasn't aware that existed